### DISCUSSION TITLE
Handling Azure Active Directory login - can't handle "Stay signed in" window with cy.origin

### DISCUSSION THREAD

**Discussion by Senpuria** [cypress-io/cypress#28042] [Questions and Help]
Hi all,

I've ran into a bit of trouble using cy.origin trying to login to Azure Active Directory (login.microsoftonline.com). I've been following the docs here https://docs.cypress.io/guides/end-to-end-testing/azure-active-directory-authentication however the docs don't go through handling the "stay signed in" window that could appear after entering a Username and Password.

This is fine, since it's just an extra click, however after following the docs, I CANNOT get it to perform an extra click after submitting the password.

The function is as follows:

`function loginViaAAD(username: string, password: string) {
    cy.visit('/')
  
    // Login to your AAD tenant.
    cy.origin(
      'login.microsoftonline.com',
      {
        args: {
          username,
          password,
        },
      },
      ({ username, password }) => {
        cy.get('input[type="email"]').type(username, {
          log: false,
        }),
        cy.get('input[type="submit"]').click()

        cy.get('input[type="password"]').type(password, {
          log: false,
        }),
        cy.get('input[type="submit"]').click()
        *****WE NEVER HIT THIS*****
        cy.get('input[type="submit"]').click()
      }
    )
    
    cy.url().should('equal', '/')
    cy.get('#welcome-div').should(
    'contain',
    `Welcome ${Cypress.env('aad_username')}!`
  )
}`

As you can see, quite simple and essentially copy and pasted from the cypress docs. Apart from we never hit that second click, cypress is expecting a new URL which I would assume it is expecting it to be the original URL for our product.

Am I doing something wrong here? I would expect to be able to perform as many actions as I want before it jumps out of the function but it doesn't seem to be doing so.

Thanks for any help!

**Comment by portyu9** [comment]
Hi @Senpuria 

`cy.origin()` is not limited to two actions. The issue is that the Microsoft authentication flow can **change origin after the email step**, so keeping email, password, and the "Stay signed in?" action inside one `login.microsoftonline.com` callback is not reliable.

The current Cypress AAD guide explicitly accounts for this. It handles the username on:

```js
cy.origin('login.microsoftonline.com', ...)
```

and notes that, depending on the Microsoft account, the password flow may continue on:

```js
cy.origin('login.live.com', ...)
```

The current example then handles Microsoft's post-password prompt inside that second origin as well.

So first check the URL that Cypress reaches after the username submit. If it changes to `login.live.com`, split the flow into separate `cy.origin()` blocks instead of trying to perform everything inside the original `login.microsoftonline.com` callback.

Also keep the currently required E2E setting:

```js
e2e: {
  experimentalModifyObstructiveThirdPartyCode: true,
}
```

and use `cy.session()` to cache the completed login, as the current AAD guide recommends.

One unrelated correction in the example:

```js
cy.url().should('equal', '/')
```

cannot succeed because `cy.url()` yields the complete URL. For just the path, use:

```js
cy.location('pathname').should('eq', '/')
```

The important diagnostic here is therefore the **actual Microsoft origin after each submit**, not the number of Cypress commands inside `cy.origin()`.

Please mark this answer as accepted if it helped, thank you.
